design/events.md

```diff
@@ -0,0 +1,439 @@
+The `EventListener` interface is where events are handed over to subscribers. Each event type has a distinct method.
+Here is a tentative class diagram:
+
+![Event listener](./events/event_listener.png)
+
+### 2.2.3. De-registration
+
+De-registration of a previously-registered subscriber is out of scope.
+
+## 2.3. Event delivery
+
+Events should be emitted _after the call to the version store, and only if the call succeeded_. This is indeed the best
+level in the call stack to intercept events. For example, a commit event could be emitted after the call to
+`org.projectnessie.versioned.VersionStore.commit()`. The delegate pattern is already used at this level to inject
+telemetry, cf. `TracingVersionStore`. A similar delegate pattern could be used for events.
+
+Note: the new storage model has the ability to recover from a failed write upon the next read targeting the same
+reference. Whether the corresponding event should be triggered on the failed write or on the recovery read is left to be
+sorted out later.
+
+### 2.3.1. Asynchronous delivery
+
+**Event delivery implementation must not block a Nessie API call, nor jeopardize the server's stability and
+responsiveness.** Instead, events will require asynchronous processing, and likely, an event emission queue.
+Implementation details will be sorted out later, including mitigation actions when the event queue gets full.
+
+### 2.3.2. Ordering guarantees
+
+The system does not provide any ordering guarantees. Events may be emitted in an out-of-sequence manner.
+
+### 2.3.3. Delivery contract
+
+Event delivery is the act of handing over an event to the subscriber's `EventListener`. For optimal event delivery,
+implementers of the `EventListener` interface are expected to respect the following requirements:
+
+* Methods in the `EventListener` interface SHOULD NOT block the calling thread. If required, implementers of the
+  `EventListener` interface should rather dispatch the calls to separate thread pools, if the underlying event
+  processing requires blocking I/O.
+* Methods in the `EventListener` interface operate in fire and forget mode: the method SHOULD return as soon as the
+  event was published to the destination queue, not when it was read by all consumers. In other words, implementations
+  should not wait for read acknowledgements.
+* The server COULD impose a (configurable) timeout after which the thread calling the `EventListener` method is
+  interrupted and delivery is considered failed.
+* A `RuntimeException` MAY be thrown to indicate that event delivery wasn't possible, e.g. because the underlying
+  pub/sub system is unreachable. But in this case, the exception SHOULD be thrown immediately.
+* For increased resilience, a failed invocation of the subscriber's `EventListener` method MAY be retried a configurable
+  number of times, with exponential backoff.
+
+**This effectively makes the system provide at-most-once delivery guarantees**. Subscribers must be prepared to miss
```

Comment:
Retries may result in more-than-once though. Maybe "mostly once" or "ideally once"?

Reply:
Best-effort? I'll go with "best-effort" :-D